File: src/game/population.js
import { noteEvidence } from "./story.js";
export const MAX_POPULATION = 1e15;
export function syncPopulation(w) {
  const room = MAX_POPULATION - w.creatures.length;
  w.cohort = Math.max(0, Math.min(room, Math.floor(w.cohort)));
  w.orbital.population = Math.max(0,
    Math.min(room - w.cohort, Math.floor(w.orbital.population)));
  w.population = w.creatures.length + w.cohort + w.orbital.population;
}
export function launch(w, c, remember) {
  if (w.creatures.length <= 8) return false;
  if (c.carry) {
    w.inventory[c.cargoKind || "wood"] += c.carry;
    c.carry = 0;
  }
  w.creatures = w.creatures.filter((o) => o.id !== c.id);
  w.orbital.population++;
  w.orbital.launches++;
  w.orbital.lastLaunch = w.time;
  const cannon = w.objects.find((o) => o.type === "cannon");
  if (cannon) cannon.phase = w.time;
  w.departed.push({ id: c.id, name: c.name, cause: "orbit", tick: w.time });
  w.departed = w.departed.slice(-32);
  noteEvidence(w, "launches", `${c.name} left for orbit.`, 1, c.id);
  remember(w, "launch", `${c.name} is part of the orbital collective.`, c.id);
  syncPopulation(w);
  return true;
}
export function stepCohorts(w, dt) {
  const orbital = w.orbital;
  orbital.fraction = 0;
  if (orbital.population > 0 && w.stage === 2) {
    // Residents here are actual arrivals. Time passing must not add people.
    const energy = orbital.population * 0.025 * dt;
    w.progress.energy = Math.min(MAX_POPULATION, w.progress.energy + energy);
    orbital.energy = Math.min(MAX_POPULATION, orbital.energy + energy);
  }
  const homes = w.objects
    .filter((o) => o.type === "dwelling")
    .reduce((s, o) => s + o.level * 48, 0);
  w.district.capacity = homes;
  w.district.population = w.cohort;
  if (w.cohort > 0) {
    const supported = Math.min(w.cohort, homes),
      unsupported = Math.max(0, w.cohort - supported);
    w.district.health = Math.max(
      0,
      Math.min(100, w.district.health + dt * (unsupported ? -0.15 : 0.2)),
    );
    const growth = supported * 0.003 * dt + w.district.fraction,
      births = Math.min(Math.floor(growth), Math.max(0, homes - w.cohort),
        // Keep district births within the shared census capacity.
        Math.max(0, MAX_POPULATION - w.creatures.length - w.cohort - orbital.population));
    w.district.fraction = growth - Math.floor(growth);
    w.cohort += births;
    if (unsupported && w.district.health === 0) {
      const lost = Math.min(
        unsupported,
        Math.floor(unsupported * 0.005 * dt + w.district.lossFraction),
      );
      w.district.lossFraction =
        (unsupported * 0.005 * dt + w.district.lossFraction) % 1;
      w.cohort -= lost;
      if (lost) {
        noteEvidence(
          w,
          "neglect",
          `${lost} unrepresented residents were lost without housing.`,
          lost,
        );
        noteEvidence(
          w,
          "deaths",
          `${lost} district residents died from unmet needs.`,
          lost,
        );
      }
    }
  }
  syncPopulation(w);
  refreshDistrict(w);
}
export function refreshDistrict(w) {
  const homes = w.objects
    .filter((o) => o.type === "dwelling")
    .reduce((s, o) => s + o.level * 48, 0);
  w.district.capacity = homes;
  const supported = Math.min(w.cohort, homes),
    unsupported = Math.max(0, w.cohort - supported);
  w.district.population = w.cohort;
  w.district.healthCounts = {
    healthy: supported,
    unwell: w.district.health > 0 ? unsupported : 0,
    critical: w.district.health === 0 ? unsupported : 0,
  };
  w.district.allocations = {
    care: Math.ceil(supported / 8),
    rest: Math.max(0, w.cohort - Math.ceil(supported / 8)),
  };
  w.district.rates = {
    births: w.cohort < homes ? supported * 0.003 : 0,
    deaths: w.district.health === 0 ? unsupported * 0.005 : 0,
    energy: 0,
  };
}

File: engine/src/population.rs
use crate::{Engine, value::*};
use serde_json::{Value, json};
pub const MAX_POPULATION: f64 = 1e15;
impl Engine {
    pub(crate) fn sync_population(&mut self) {
        let count = list(&self.world, "creatures").len() as f64;
        let room = MAX_POPULATION - count;
        self.world["cohort"] = json!(num(&self.world, "cohort").floor().min(room).max(0.));
        self.world["orbital"]["population"] = json!(
            num(&self.world["orbital"], "population")
                .floor()
                .min(room - num(&self.world, "cohort"))
                .max(0.)
        );
        self.world["population"] =
            json!(count + num(&self.world, "cohort") + num(&self.world["orbital"], "population"));
    }
    pub(crate) fn launch(&mut self, c: &mut Value) -> bool {
        if list(&self.world, "creatures").len() <= 8 {
            return false;
        }
        if num(c, "carry") > 0. {
            let kind = if text(c, "cargoKind").is_empty() {
                "wood"
            } else {
                text(c, "cargoKind")
            }
            .to_string();
            increment(&mut self.world["inventory"], &kind, num(c, "carry"));
            c["carry"] = json!(0);
        }
        self.density_index.borrow_mut().remember_position(c);
        self.world["creatures"]
            .as_array_mut()
            .unwrap()
            .retain(|o| !same_id(&o["id"], &c["id"]));
        increment(&mut self.world["orbital"], "population", 1.);
        increment(&mut self.world["orbital"], "launches", 1.);
        let time = num(&self.world, "time");
        self.world["orbital"]["lastLaunch"] = json!(time);
        if let Some(o) = self.world["objects"]
            .as_array_mut()
            .unwrap()
            .iter_mut()
            .find(|o| text(o, "type") == "cannon")
        {
            o["phase"] = json!(time);
        }
        let departed = self.world["departed"].as_array_mut().unwrap();
        departed.push(json!({"id":c["id"],"name":c["name"],"cause":"orbit","tick":time}));
        if departed.len() > 32 {
            departed.drain(..departed.len() - 32);
        }
        self.note_evidence(
            "launches",
            &format!("{} left for orbit.", text(c, "name")),
            1.,
            Some(text(c, "id")),
        );
        self.remember(
            "launch",
            &format!("{} is part of the orbital collective.", text(c, "name")),
            Some(text(c, "id")),
        );
        self.sync_population();
        true
    }
    pub(crate) fn step_cohorts(&mut self, dt: f64) {
        self.world["orbital"]["fraction"] = json!(0);
        if num(&self.world["orbital"], "population") > 0. && num(&self.world, "stage") == 2. {
            let energy = num(&self.world["orbital"], "population") * 0.025 * dt;
            self.world["progress"]["energy"] =
                json!((num(&self.world["progress"], "energy") + energy).min(MAX_POPULATION));
            self.world["orbital"]["energy"] =
                json!((num(&self.world["orbital"], "energy") + energy).min(MAX_POPULATION));
        }
        let homes = self.housing_capacity();
        self.world["district"]["capacity"] = json!(homes);
        self.world["district"]["population"] = self.world["cohort"].clone();
        let cohort = num(&self.world, "cohort");
        if cohort > 0. {
            let supported = cohort.min(homes);
            let unsupported = (cohort - supported).max(0.);
            self.world["district"]["health"] = json!(
                (num(&self.world["district"], "health")
                    + dt * if unsupported > 0. { -0.15 } else { 0.2 })
                .clamp(0., 100.)
            );
            let growth = supported * 0.003 * dt + num(&self.world["district"], "fraction");
            let births = growth.floor().min((homes - cohort).max(0.)).min(
                (MAX_POPULATION
                    - list(&self.world, "creatures").len() as f64
                    - cohort
                    - num(&self.world["orbital"], "population"))
                .max(0.),
            );
            self.world["district"]["fraction"] = json!(growth - growth.floor());
            increment(&mut self.world, "cohort", births);
            if unsupported > 0. && num(&self.world["district"], "health") == 0. {
                let amount =
                    unsupported * 0.005 * dt + num(&self.world["district"], "lossFraction");
                let lost = unsupported.min(amount.floor());
                self.world["district"]["lossFraction"] = json!(amount % 1.);
                increment(&mut self.world, "cohort", -lost);
                if lost > 0. {
                    self.note_evidence(
                        "neglect",
                        &format!(
                            "{} unrepresented residents were lost without housing.",
                            lost
                        ),
                        lost,
                        None,
                    );
                    self.note_evidence(
                        "deaths",
                        &format!("{} district residents died from unmet needs.", lost),
                        lost,
                        None,
                    );
                }
            }
        }
        self.sync_population();
        self.refresh_district();
    }
    fn housing_capacity(&self) -> f64 {
        list(&self.world, "objects")
            .iter()
            .filter(|o| text(o, "type") == "dwelling")
            .map(|o| num(o, "level") * 48.)
            .sum()
    }
    pub(crate) fn refresh_district(&mut self) {
        let homes = self.housing_capacity();
        let cohort = num(&self.world, "cohort");
        let supported = cohort.min(homes);
        let unsupported = (cohort - supported).max(0.);
        let healthy = num(&self.world["district"], "health") > 0.;
        self.world["district"]["capacity"] = json!(homes);
        self.world["district"]["population"] = json!(cohort);
        self.world["district"]["healthCounts"] = json!({"healthy":supported,"unwell":if healthy{unsupported}else{0.},"critical":if healthy{0.}else{unsupported}});
        self.world["district"]["allocations"] =
            json!({"care":(supported/8.).ceil(),"rest":(cohort-(supported/8.).ceil()).max(0.)});
        self.world["district"]["rates"] = json!({"births":if cohort<homes{supported*0.003}else{0.},"deaths":if healthy{0.}else{unsupported*0.005},"energy":0});
    }
}
